### PR TITLE
windows: fix fs-promises-writeFile-async-iterator.test.ts [v2]

### PR DESCRIPTION
**Pull request by nektro**
Closes https://github.com/oven-sh/bun/issues/13152

adds independent tests too

### CODE REVIEW

**Comment by robobun** [comment]
@nektro, your commit 2e73560a0a9f12b505a8932c20fbba12e140061c is building: [`#1518`](https://buildkite.com/bun/bun/builds/1518)

File: src/io/PipeWriter.zig [hunk 1]
```diff
@@ -1003,7 +1003,12 @@ pub fn WindowsBufferedWriter(
             this.is_done = true;
             if (this.pending_payload_size == 0) {
                 // will auto close when pending stuff get written
-                this.close();
+                if (this.source) |source| {
+                    switch (source) {
+                        .sync_file, .file => {},
```
Comment (Jarred-Sumner) [inline src/io/PipeWriter.zig]: this is confusing to me

Comment (Jarred-Sumner) [reply src/io/PipeWriter.zig]: can you add a comment explaining why it should only call close for pipes and ttys?

File: src/io/PipeWriter.zig [hunk 2]
```diff
@@ -1382,7 +1387,12 @@ pub fn WindowsStreamingWriter(
             this.closed_without_reporting = false;
             // if we are done we can call close if not we wait all the data to be flushed
             if (this.isDone()) {
-                this.close();
+                if (this.source) |source| {
+                    switch (source) {
+                        .sync_file, .file => {},
```
Comment (Jarred-Sumner) [inline src/io/PipeWriter.zig]: Please add a comment explaining this

File: test/js/bun/util/filesink.test.ts [hunk 3]
```diff
@@ -146,3 +146,31 @@ describe("FileSink", () => {
     });
   }
 });
+
+import fs from "node:fs";
+import util from "node:util";
+import path from "node:path";
+
+it("end doesn't close when backed by a file descriptor", async () => {
```
Comment (Jarred-Sumner) [inline test/js/bun/util/filesink.test.ts:173]: Where does it check the opposite - that `Bun.file(path).writer().end()` does close the file descriptor?

Comment (nektro) [reply test/js/bun/util/filesink.test.ts:173]: this fails on posix as well. should it pass?
```ts
it("end does close when not backed by a file descriptor", async () => {
  const x = tmpdirSync();
  const file = Bun.file(path.join(x, "test.txt"));
  const writer = file.writer();
  await writer.write(Buffer.from("1 Hello, world!"));
  await writer.end();
  expect(async () => await writer.write(Buffer.from("2 Hello, world!"))).toThrow();
});
```

Comment (nektro) [reply test/js/bun/util/filesink.test.ts:173]: okay, using `getMaxFD()` instead and `expect(fd_after).toBe(fd_before);` does pass on posix

Comment (nektro) [reply test/js/bun/util/filesink.test.ts:173]: and does not on windows, i'll use that check instead

File: src/io/PipeWriter.zig [hunk 4]
```diff
@@ -891,6 +891,7 @@ pub fn WindowsBufferedWriter(
 ) type {
     return struct {
         source: ?Source = null,
+        is_fd: ?bool = null,
```
Comment (Jarred-Sumner) [inline src/io/PipeWriter.zig]: this variable name doesn't say what it is used for

it's also an optional bool which takes up extra space

instead, couldn't this be an enum key on `Source`? like `owned_file` to say "a file which should be closed"

**Pull request merged**